libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+  backend driver for airspeed from www.qio-tek.com a I2C ASP5033 sensor
+ */
+
+#include "AP_Airspeed_ASP5033.h"
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <AP_Math/AP_Math.h>
+
+extern const AP_HAL::HAL &hal;
+
+#define ASP5033_I2C_ADDR 0x6D
+#define SYS_CONFIG       0xA5
+#define CMD              0x30
+#define CMD_MEASURE      0x0A
+#define PRESS_DATA       0x06
+#define TEMP_DATA        0x09
+#define WAI_REG          0x01
+#define WAI_SET          0XA4
+#define DEVICE_ID        0x66
+#define SENSOR_READY     0x08
+
+
+AP_Airspeed_ASP5033::AP_Airspeed_ASP5033(AP_Airspeed &_frontend, uint8_t _instance) :
+    AP_Airspeed_Backend(_frontend, _instance)
+{
+}
+
+// probe and initialise the sensor
+bool AP_Airspeed_ASP5033::init()
+{
+    uint8_t SysID;
+    uint8_t whoami;
+    _dev = hal.i2c_mgr->get_device(get_bus(), ASP5033_I2C_ADDR);
+    if (!_dev) {
+        return false;
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"ASP5033: sensor not found\n");
+    }
+    _dev->get_semaphore()->take_blocking();
+    _dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+    _dev->write_register(WAI_SET, DEVICE_ID); //Set Whoami ID. 0XA4 is a OPT WhoamI Part ID register, also can be read from address 0x01.
+    //check whoami, set the PartID register to 0X66 first, them Read the PartID register.
+    if (!_dev->read_registers(WAI_REG, &whoami, 1) || //Read the WhoamI ID
+        whoami != DEVICE_ID){
+        _dev->get_semaphore()->give();
+        return false;
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"ASP5033: sensor not found\n"); 
+    }
+    _dev->set_retries(2);
```

Comment:
Presumably you want to do the retries before reading the whoami, just incase there's an error. It also simplifies the semaphore you are holding, as you can just give it once right after setting it, rather then having to remember to give it on different code paths..